TESTS/mbed_hal/sleep_manager/main.cpp

```diff
@@ -16,45 +16,169 @@
+        lock_count--;
+        TEST_ASSERT_FALSE(sleep_manager_can_deep_sleep());
+    }
+    sleep_manager_unlock_deep_sleep();
+    TEST_ASSERT_TRUE(sleep_manager_can_deep_sleep());
+}
+
+void test_lock_gt_ushrt_max()
+{
+    uint32_t lock_count = 0;
+    while (lock_count < USHRT_MAX) {
+        sleep_manager_lock_deep_sleep();
+        lock_count++;
+        TEST_ASSERT_FALSE(sleep_manager_can_deep_sleep());
+    }
+
+    uint32_t expected_err_count = num_test_errors + 1;
+    sleep_manager_lock_deep_sleep();
+    TEST_ASSERT_EQUAL_UINT32(expected_err_count, num_test_errors);
+
+    // Make sure upcoming tests won't be broken.
+    while (lock_count > 0) {
+        sleep_manager_unlock_deep_sleep();
+        lock_count--;
+    }
+    TEST_ASSERT_TRUE(sleep_manager_can_deep_sleep());
 }
 
-utest::v1::status_t greentea_test_setup(const size_t number_of_cases) 
+#if DEVICE_LPTICKER
+void wakeup_callback(volatile int* wakeup_flag)
 {
-    GREENTEA_SETUP(20, "default_auto");
-    return greentea_test_setup_handler(number_of_cases);
+    (*wakeup_flag)++;
+}
+
+void test_sleep_auto()
+{
+    const ticker_data_t* const us_ticker = get_us_ticker_data();
+    const ticker_data_t* const lp_ticker = get_lp_ticker_data();
+    us_timestamp_t us_ts, lp_ts, us_diff1, us_diff2, lp_diff1, lp_diff2;
+    LowPowerTimeout lp_timeout;
+
+    sleep_manager_lock_deep_sleep();
+    volatile int wakeup_flag = 0;
+    lp_timeout.attach_us(mbed::callback(wakeup_callback, &wakeup_flag), SLEEP_DURATION_US);
+    us_ts = ticker_read_us(us_ticker);
+    lp_ts = ticker_read_us(lp_ticker);
+
+    while (wakeup_flag == 0) {
+        sleep_manager_sleep_auto();
+    }
+    us_diff1 = ticker_read_us(us_ticker) - us_ts;
+    lp_diff1 = ticker_read_us(lp_ticker) - lp_ts;
+    // Verify that us and lp time differences are equal with 10% tolerance.
+    TEST_ASSERT_UINT64_WITHIN_MESSAGE( \
+            SLEEP_DURATION_US / 10ULL, lp_diff1, us_diff1, \
+            "us ticker peripheral should be powered on when deep sleep mode us locked");
+
+    sleep_manager_unlock_deep_sleep();
+    TEST_ASSERT_TRUE(sleep_manager_can_deep_sleep());
+
+    // Wait 20ms for hardware serial buffers to flush.
+    wait_ms(20);
+
+    wakeup_flag = 0;
+    lp_timeout.attach_us(mbed::callback(wakeup_callback, &wakeup_flag), SLEEP_DURATION_US);
+    us_ts = ticker_read_us(us_ticker);
+    lp_ts = ticker_read_us(lp_ticker);
+
+    while (wakeup_flag == 0) {
+        sleep_manager_sleep_auto();
+    }
+    us_diff2 = ticker_read_us(us_ticker) - us_ts;
+    lp_diff2 = ticker_read_us(lp_ticker) - lp_ts;
+
+    TEST_ASSERT_MESSAGE(us_diff2 < lp_diff2 / 4, "Deep sleep mode not used when unlocked");
+    TEST_ASSERT_MESSAGE(us_diff2 < us_diff1 / 4, "Deep sleep mode not used when unlocked");
+}
+#endif
+
+utest::v1::status_t test_setup(const size_t number_of_cases)
+{
+    GREENTEA_SETUP(10, "default_auto");
+    // Suspend the RTOS kernel scheduler to prevent any OS interference.
+    osKernelSuspend();
```

Comment:
NOTE:
As @mprse pointed out, when https://github.com/ARMmbed/mbed-os/pull/7524 is merged, below code should be added here.
```cpp
#if DEVICE_LPTICKER && (LPTICKER_DELAY_TICKS > 0)
    // After the OS is disabled wait until the microsecond timer
    // is no longer in use by the lp ticker wrapper code.
    // This prevents the low power ticker wrapper code from
    // getting stuck in a state where it is waiting for
    // the microsecond Timeout to run.
    while (lp_ticker_get_timeout_pending());
#endif
```